[영화 극장 선택] 팡태 1, 2단계 미션 제출합니다

Hevton opened the pull request:
안녕하세요 제이든! 또 만나게 되어 반갑습니다!
이번엔 첫 3인 페어로 진행했는데, 각자의 다양한 끼를 맛볼 수 있어서 좋았습니다.
페어 [하디의 코드](https://github.com/woowacourse/android-movie-ticket/pull/87)를 토대로 미션을 시작했습니다

[e576963](https://github.com/woowacourse/android-movie-theater/commit/e576963d3f3b7b3a8b4cd20a111d88d1cec749ca) 부터 봐주시면 감사하겠습니다!

https://github.com/woowacourse/android-movie-theater/assets/86276828/064c6ca7-caad-44da-8132-dc2c283793ce

woowahan-pjs committed:
:tada: Theater시작!

Commit:
:sparkles: 영화 선택 후 극장을 선택할 수 있음,
극장 이름 줄이기

Commit:
:recycle: 분기처리

Commit:
:recycle: dataBinding 적용

Commit:
:truck: seat->activity_theater_seat

Commit:
:truck: seat->activity_theater_seat

Commit:
:recycle: Test오류 고침

Commit:
:sparkles: 영화 시간 정보 최종 화면에 표시

murjune committed:
build: resources 파일 exclude 할 목록 추가

murjune committed:
ui: 바텀 네비 menu

murjune committed:
ui: 예약, 영화 목록, 세팅 Fragment

murjune committed:
ui: Fragment transaction 적용

murjune committed:
feat: MovieListFragment 로 마이그레이션

Commit:
:bug: 카운팅 안되던 버그 고침

murjune committed:
:recycle:  findFragmentByTag 로 개선

murjune committed:
:bug: ticket count 로직 개선

murjune committed:
:memo: 기능 요구사항 명세서 업데이트

murjune committed:
:white_check_mark: MovieListFragment test

murjune committed:
:truck: 이름 변경

murjune committed:
:truck: xml 이름 통일성

murjune committed:
:sparkles: timeSpinner 개선

murjune committed:
:recycle: TheaterViewHolder dataBinding 적용

murjune committed:
:recycle: MovieDetailActivity dataBinding으로 개선

murjune committed:
:recyle: seat activity databinding apply

murjune committed:
:recyle: Detail layout Spinner 위치 조정

murjune committed:
:recycle: :white_check_mark: PurchaseConfirmationActivityTest 구조 변경

- 액티비티 변경에 따른 테스트 코드 변경

murjune committed:
:recycle: :white_check_mark: MovieListFragmentTest 구조 변경

murjune committed:
:recycle: :white_check_box: 테스트 코드 고침

murjune committed:
:recycle: ktLint Check

Commit:
:truck: 테스트 코드 패키지 프로덕트와 동일하게 맞춤

Commit:
:memo: README.md 할 일 전부 체크 완료

Commit:
:recycle: 프레젠터나 뷰가 값을 반환하지 않음

Commit:
:recycle: 네임스페이스는 모든 태그의 최상단에 작성

KwonDae commented on README.md:
<img width="397" alt="image" src="https://github.com/woowacourse/android-movie-theater/assets/33443660/51cc3f5e-f630-44f8-9eee-7d44adea069d">

👍

Hevton replied:
제이든! 제출 전에 ktlint 와 clean test 도 진행했는데, 이 No cast needed는 어떻게 확인할 수 있는지 여쭤봐도 될까요??

KwonDae commented on app/src/main/java/woowacourse/movie/movieList/MovieListFragment.kt:
```diff
@@ -0,0 +1,87 @@
+    }
+
+    override fun updateAdapter(displayData: List<MovieDisplayData>) {
+        adapter.updateItems(displayData)
```
초기 아이템을 셋팅하는 updateItems는 잘 동작하겠지만,
함수 네이밍에 맞게끔 item을 update하는 동작이 정상적으로 되고 있을까요?

Hevton replied:
데이터 셋을 변경한 뒤에 대표적으로는 notifyDataSetChanged()를 진행해줘야 생각했는데, 이 부분이 코드에 누락되어 있었음에도 불구하고 데이터 반영과 업데이트가 계속해서 잘 되는 것을 페어들과 추가적으로 확인하면서 멘붕이 왔습니다.. 혹시 이 상황에 대해 알고 계시는 부분이 있으신가요..??

KwonDae replied:
ㅎㅎ..  어댑터가 초기화되고 RecyclerView에 연결된 이후에 어떻게 데이터를 바꿔주었길래 업데이트가 잘 되었는지 함께 적어주셨으면 상황을 잘 알 수 있겠죠..?

일단, 
```kotlin
    private fun initViews() {
        binding.rvMovies.layoutManager = LinearLayoutManager(requireContext())
        adapter =
            MovieAdapter(requireContext(), mutableListOf()) { position ->
                adapter.updateItems(emptyList())
//             adapter.notifyDataSetChanged() 를 추가해주면 반영이 잘 된다.. 왜그럴까?
            }
        binding.rvMovies.adapter = adapter
    }
```

`지금 예매` 버튼을 눌렀을 떄 어댑터에 빈 리스트로 업데이트 해준다고 했을 때 반영이 잘 안되고 있으며, 당연한 결과입니다.

KwonDae replied:
앞으로의 미션이나, 나중에 실무에 가셔서도 질문을 잘 하는법을 익히셔야 합니다.

1. 현재 나는 어디까지 시도해보았고
2. 어떻게 구현을 해보았으며
3. 어떠한 부분에서 어려움을 겪고있는지

정도의 정보를 전달해야 동료 개발자와의 원활한 커뮤니케이션이 가능합니다.
https://brunch.co.kr/@codestates/4

Hevton replied:
정말 죄송합니다. PR 메세지를 보내지 않아도 알림이 가는군요.. 
테스트를 마저 정리하여 코멘트를 구체화 한 뒤에 리뷰 요청을 보내려고 했는데, 확인하실 줄을 몰랐습니다. 
다음부터 정말 주의하겠습니다..

RecyclerView나 ListView의 어댑터의 변경사항이 반영되기 위해서는 notifyDataSetChanged()를 진행해야 하는 것으로 알고 있습니다.

그런데 생각을 해보니 앱을 실행시킨 맨 처음에는 notifyDataSetChanged()를 호출하지 않아도 
updateAdapter()의 인자로 넘어온 변경사항 데이터가 왜 반영이 되는지 의문이 들었습니다.
```kotlin
class MovieAdapter(
    private val context: Context,
    private val onClick: (Int) -> Unit,
) : RecyclerView.Adapter<RecyclerView.ViewHolder>() {
    private var items: List<MovieDisplayData> = emptyList()
}
```
```kotlin
// MovieListFragment
    override fun updateAdapter(displayData: List<MovieDisplayData>) {
        adapter.updateItems(displayData)
    }
```
 
그에 대한 테스트를 다양하게 진행해 본 결과 다음과 같은 추론을 하게 되었습니다.

```kotlin
  override fun updateAdapter(displayData: List<MovieDisplayData>) {
        adapter.updateItems(emptyList()) // 뷰에 반영 안됨
        adapter.updateItems(displayData) // 뷰에 반영됨
    }

```
```kotlin
    fun updateAdapter2(displayData: List<MovieDisplayData>) {
        adapter.updateItems(emptyList()) // 뷰에 반영됨
        lifecycleScope.launch {
            delay(100)
            adapter.updateItems(displayData) // 뷰에 반영 안됨
        }
    }

```
RecylcerView가 자신을 보여줄 준비를 할 때, 어댑터의 getItemCount()을 호출하게 될텐데, 만약 이 함수가 호출되기 전에 데이터가 업데이트 되면 notifyDataSetChanged()를 하지 않아도 반영이 되는 것인가? 라는 생각이 들었습니다.

KwonDae replied:
> 그런데 생각을 해보니 앱을 실행시킨 맨 처음에는 notifyDataSetChanged()를 호출하지 않아도
> updateAdapter()의 인자로 넘어온 변경사항 데이터가 왜 반영이 되는지 의문이 들었습니다.

그 이유는 MovieListFragment에서
- initViews()
- updateAdapter
- Recyclerview 내부 getItemCount 

순으로 호출되기 때문에 notifyDataSetChanged 함수 없이도 반영이 되는 것 처럼 보였습니다.
위 순서 확인은 아래 디버깅 방법을 통해 알 수 있습니다.

지금은 운이 좋게 반영이 되었지만, updateAdapter 함수가 어떠한 이유로 호출이 늦게 된다면,
가져온 데이터를 표시할 수 있을까요?

KwonDae replied:
> 그런데 생각을 해보니 앱을 실행시킨 맨 처음에는 notifyDataSetChanged()를 호출하지 않아도
> updateAdapter()의 인자로 넘어온 변경사항 데이터가 왜 반영이 되는지 의문이 들었습니다.

그 이유는 MovieListFragment에서
- initViews()
- updateAdapter
- Recyclerview 내부 getItemCount 

순으로 호출되기 때문에 notifyDataSetChanged 함수 없이도 반영이 되는 것 처럼 보였습니다.
위 순서 확인은 아래 디버깅 방법을 통해 알 수 있습니다.

지금은 운이 좋게 반영이 되었지만, updateAdapter 함수가 어떠한 이유로 호출이 늦게 된다면,
가져온 데이터를 표시할 수 있을까요?

Hevton replied:
감사합니다! 저도 동의합니다! 실제로 위처럼 delay를 100 주고 나니까 반영이 되지 않더라구요 ㅎㅎ..
함께 이야기 나눠주셔서 감사합니다!

KwonDae commented on app/src/main/java/woowacourse/movie/movieList/MovieListFragment.kt:
```diff
@@ -0,0 +1,87 @@
+    }
+
+    override fun showBottomSheet(theater: Theater) {
+        val fragment =
+            (
+                childFragmentManager.findFragmentByTag(TheatersBottomSheetFragment.TAG)
+                    as? TheatersBottomSheetFragment
+            ) ?: TheatersBottomSheetFragment()
```
이 로직에 대해서 설명을 해주실 수 있으신가요?
childFragmentManager 는 어떤 역할을 하고있나요?

Hevton replied:
TAG를 통해서 찾아진다면 재활용해주고, 아니면 새로 생성하고자 하는 의도로 작성했습니다!
그리고 MovieListFragment에서 띄우는 BottomSheetFragment이므로, 자기 자신의 프래그먼트 매니저인 childFragmentManager를 사용하여 parentFragmentManager까지 가지 않아도 된다고 생각했습니다!

KwonDae replied:
```kotlin
    override fun showBottomSheet(theater: Theater) {
        var fragment = childFragmentManager.findFragmentByTag(TheatersBottomSheetFragment.TAG) as? TheatersBottomSheetFragment

        if (fragment == null) {
            fragment = TheatersBottomSheetFragment {
// 바텀시트 내 아이템을 클릭했을 때 showBottomSheet를 다시 호출하게 함으로써 fragment가 재활용이 되는지 체크하기 위함
                showBottomSheet(theater)
            }
        }
        val bottomSheet =
            fragment.withArgs {
                putSerializable(THEATER_KEY, theater)
            }
        bottomSheet.show(childFragmentManager, TheatersBottomSheetFragment.TAG)
    }
```

fragment의 객체 값의 디버깅 테스트를 위해 위와 같이 코드를 변경하고

<img width="1077" alt="image" src="https://github.com/woowacourse/android-movie-theater/assets/33443660/cc8fc107-659d-486b-b650-22ef30eff220">

https://developer.android.com/codelabs/basic-android-kotlin-compose-intro-debugger?hl=ko#0

위 링크를 참조하여 안드로이드에서는 디버깅을 어떻게 할 수 있는지 확인해보셔서 직접 fragment 값을 확인해보시길 바랍니다.
**왜 재활용이 되지 않는지는**

힌트는 `findFragmentByTag` 와 BottomSheetFragment 의 `dismiss` 함수의 작동방식을 살펴보세요.

Hevton replied:
사진에서처럼 dissmiss할때 Fragment가 popBackStackImmediate() 되어 재활용이 안되므로
![image](https://github.com/woowacourse/android-movie-theater/assets/86276828/b137e57e-c1a4-49a4-b52f-476c00c25a8a)

그냥 사용해도 된다고 느꼈습니다..! 감사합니다!

KwonDae commented on app/src/main/java/woowacourse/movie/movieList/MovieListFragment.kt:
```diff
@@ -0,0 +1,87 @@
+    private fun initViews() {
+        binding.rvMovies.layoutManager = LinearLayoutManager(requireContext())
+        adapter =
+            MovieAdapter(requireContext(), mutableListOf()) { position ->
```
Adapter에 가변컬렉션을 넘겨야 하는 이유가 있는걸까요? 👀

Hevton replied:
이 부분 불필요하다고 생각되어 수정하였고, MovieAdapter 내부에서 필드로 초기화하게끔 수정했습니다! ㅎㅎ

KwonDae commented on app/src/main/java/woowacourse/movie/movieList/MovieListFragment.kt:
```diff
@@ -0,0 +1,87 @@
+    }
+
+    private fun initViews() {
+        binding.rvMovies.layoutManager = LinearLayoutManager(requireContext())
```
xml에서 설정이 가능합니다
`app:layoutManager="androidx.recyclerview.widget.LinearLayoutManager"`

Hevton replied:
감사합니다!

KwonDae commented on app/src/main/java/woowacourse/movie/movieList/cinemaListDialog/CinemaAdapter.kt:
```diff
@@ -4,34 +4,46 @@ import android.view.LayoutInflater
+                    oldItem: Cinema,
+                    newItem: Cinema,
+                ): Boolean {
+                    return oldItem.cinemaName == newItem.cinemaName
```
ListAdapter를 활용해주셨는데요,
영화이름은 중복으로 많이 존재할 수 있는데 name으로 비교해도 괜찮을까요?

areItemsTheSame, areContentsTheSame은 각각 어떤 역할을 할까요?

Hevton replied:
id로 비교를 하면 더 좋겠지만 현재 상황에서 도메인에 id가 없어서 name으로 비교해주었는데요 ㅜㅜ
areItemsTheSame 에서 true가 되더라도 areContentsTheSame 에서 두 아이템이 같지 않다는 것이 증명될 것이기에
이 상황에서도 괜찮지 않을까?! 라는 생각이 있는데, 제이든은 어떻게 생각하시나요?

KwonDae replied:
그렇다면 name으로 비교하는게 현재로선 의미가 없어졌네요 🥲
말씀하신대로 지금은 id가 존재하지 않고 현재는 ListAdapter를 다루지 않기에
areItemsTheSame, areContentsTheSame 의 역할만 알아두고 넘어가셔도 무방합니다.

Hevton replied:
감사합니다!

KwonDae commented on app/src/main/java/woowacourse/movie/movieList/cinemaListDialog/TheatersBottomSheetFragment.kt:
```diff
@@ -11,58 +11,68 @@ import woowacourse.movie.databinding.FragmentBottomSheetTheatersBinding
+                                LocalTime.of(23, 30),
+                            ),
+                    ),
                 ),
```
View에 존재해야 하는 값들 일까요?

Hevton replied:
Presenter에서 model과 같은 곳에서 읽어온다는 내용으로 바꿨습니다!

KwonDae commented on app/src/main/java/woowacourse/movie/movieDetail/MovieDetailActivity.kt:
```diff
@@ -93,29 +75,24 @@ class MovieDetailActivity : AppCompatActivity(), MovieDetailContract.View {
+                        binding.movieDateSpinner.selectedItem.toString() + " " + binding.movieTimeSpinner.selectedItem.toString(),
+                    )
                 }
             navigateToPurchaseConfirmation(intent)
```
다양한 데이터를 넘겨야 하는 경우 실수를 유발할 수 있습니다.
타켓 Activity에서 
```kotlin
fun createIntent(context, no, name, date) {
       return Intent(context).apply....
}
```

식으로 해당 액티비티는 어떤 정보들을 요구하는지 포맷을 만들어 제공해보는건 어떨까요?

Hevton replied:
너무 좋은 방법이네요...! 감사합니다 적용했습니다!

KwonDae commented on app/src/main/java/woowacourse/movie/movieList/cinemaListDialog/TheatersBottomSheetFragment.kt:
```diff
@@ -77,7 +87,12 @@ class TheatersBottomSheetFragment : BottomSheetDialogFragment() {
                     putExtra("Cinema", cinema)
                 }
+            dismiss()
             startActivity(intent)
```
https://stackoverflow.com/questions/50825186/is-finish-always-scheduled-before-startactivity

호출 순서에 대해서도 한번 찾아보셔도 좋은 내용입니다.

Hevton replied:
알려주셔서 감사합니다!

KwonDae commented on app/src/main/java/woowacourse/movie/movieList/cinemaListDialog/TheatersBottomSheetFragment.kt:
```diff
@@ -77,7 +87,12 @@ class TheatersBottomSheetFragment : BottomSheetDialogFragment() {
                 Intent(this, MovieDetailActivity::class.java).apply {
```
여기도 마찬가지로 cinema라는 객체를 필요하다는 가이드를 줄 수 있게끔 포맷을 제공해도 좋습니다.

Hevton replied:
감사합니다!

KwonDae commented on app/src/main/java/woowacourse/movie/movieDetail/MovieDetailPresenter.kt:
```diff
@@ -1,57 +1,49 @@
-        movie?.let { view.initializeViews(it) }
+    private lateinit var movie: MovieInfo
+
+    override fun getTickets(): Int = ticketNum
+
+    override fun getMovie(): MovieInfo = movie
```
Presenter는 어떤 값도 반환하지 않습니다

Hevton replied:
궁금한 사항이 생겼습니다! MVP 패턴에서 DataBinding을 할 때 Presenter가 데이터를 갖고 있으므로 Presenter와 연결해주어야겠다고 생각했는데, Presenter와 데이터 바인딩 하는 것은 권장되지 않는다고 들었습니다. 하지만 어떤 이유에서 그런 것일지 아직 잘 모르겠습니다.. 그리고 만약 Activity나 Fragment와 바인딩해야한다면, Activity와 Fragment가 데이터를 가지고 있어야 하는데.. 그것이 MVP인가에 대해서는 아니라는 생각이 들어서, 조언을 구할 수 있을까요?

KwonDae replied:
Activity/Fragment가 값을 가지지 않고, Presenter의 값을 주입해주면 해결될 것 같은데요,
Presenter를 직접 바인딩 해주는 것을 지양하는 이유는 MVP패턴이 지향하는 방향과 맞지 않기 때문입니다.

![image](https://github.com/woowacourse/android-movie-theater/assets/33443660/0ced9b54-b191-467a-af41-7f536abf15d3)

xml은 View에 종속적인 것으로 Activity/Fragment와 상호작용을 하는 컴포넌트입니다.
각 UI 컴포넌트 마다 어떤 값을 대칭할지, 이벤트를 어떤 함수와 매칭시킬지 설정하는 곳이죠

이전 미션에서의 데이터바인딩 없이는 직접 Activity/Fragment에서 id 를 개별로 모두 찾아와 직접 onClickListener를 달아주었다면,

데이터바인딩은 이 과정을 최소화 시켜주었습니다.
Variable 을 통해 모델과 Listener를 제공받을 수 있죠.
여기서 presenter를 바로 참조한다면, presenter의 변경사항이 바로 xml 까지 영향을 끼칠 수 있는 상황이 발생합니다.
그래서 interface로 추상화한 Listener를 참조하는 이유이기도 하죠.

이부분은 MVP의 한계로 보일 수 있는 부분이지만 
View가 값을 업데이트 하도록 개선해주세요 💪

Hevton replied:
아하..! 그럼 Presenter가 View의 함수를 호출해줄 때 인자로 넘어간 모델이나 원시값들을 데이터바인딩 시키면 되겠네요!
감사합니다! 적용하겠습니다!

KwonDae commented on app/src/main/java/woowacourse/movie/movieDetail/MovieDetailPresenter.kt:
```diff
@@ -1,57 +1,49 @@
-            times.add("$time:00")
-            time += 2
-        }
+    override fun updateTimeSpinner(times: List<String>) {
```
updateTimeSpinner 라는 함수 기능은 View에 해당하는 기능 아닐까요?

프레젠터는 UI와 관련이 없도록 작성되어야 합니다.
프레젠터는 비즈니스 로직(도메인 로직)을 담당하는데 View를 업데이트 하는 함수명을 갖는게 자연스러울까요?

KwonDae replied:
View 에서 입력이벤트를 감지하고,
입력이벤트에 해당하는 비즈니스 로직을 수행하고자 domain과 소통을 해야하는데,
Model 과 View 간 의존성을 약하게 하기 위해 Presenter를 사용합니다.

Presenter는 View로 부터 해당 이벤트에 맞는 비즈니스 로직을 도메인과 소통을 하며 수행하는 장소인거죠.
따라서 비즈니스 로직 단위로 구성이 되어 있어야 합니다.

Hevton replied:
감사합니다. 함수명을 변경하여 MVP 에 가까운 설계에 대해 다시 한 번 생각해보게 되었습니다!

KwonDae commented on app/src/main/java/woowacourse/movie/movieDetail/MovieDetailPresenter.kt:
```diff
@@ -1,57 +1,49 @@
     private val view: MovieDetailContract.View,
-    private val movie: MovieInfo?,
 ) : MovieDetailContract.Presenter {
     private var ticketNum = 1
```
프레젠터가 변수값을 가지지 않고 도메인 로직 수행으로 대신할 수 없을까요?

Hevton replied:
모두 객체를 이전 Activity에서 직렬화해서 넘겨주고 받아오기에 presenter에서 데이터를 저장하게 되었습니다..!

KwonDae commented on app/src/main/java/woowacourse/movie/movieList/cinemaListDialog/TheaterViewHolder.kt:
```diff
@@ -1,17 +1,22 @@
 class TheaterViewHolder(
-    private val binding: BottomSheetTheatersItemBinding,
-    private val onCinemaClicked: (Cinema) -> Unit
+    private val binding: ItemBottomSheetTheatersBinding,
+    private val onCinemaClicked: (Cinema) -> Unit,
```
바인딩을 잘 주입받아 사용해주셨네요 👍

KwonDae commented on app/src/main/java/woowacourse/movie/movieList/cinemaListDialog/TheaterViewHolder.kt:
```diff
@@ -1,17 +1,22 @@
-        binding.tvMovieTitle.text = title
-        binding.tvMovieDescription.text = "${theater.times.size}"
+        binding.title = title
+        binding.description =
+            binding.root.context.getString(
+                R.string.movie_count_for_cinema_format,
+                theater.times.size.toString(),
+            )
```
DataBinding에서도 string format 사용이 가능합니다
https://stackoverflow.com/questions/38978499/how-do-i-use-databinding-to-combine-a-string-from-resources-with-a-dynamic-varia

Hevton replied:
감사합니다!

KwonDae commented on app/src/main/java/woowacourse/movie/seat/TheaterSeatPresenter.kt:
```diff
@@ -1,14 +1,26 @@
+        private set
+    private var totalPrice = 0
+
+    override fun getTotalPrice(): Int {
+        return totalPrice
+    }
+
+    override fun getCinema(): Cinema {
+        return cinema
+    }
```
프레젠터는 값을 반환하지 않고 비즈니스 로직을 수행할 뿐입니다
아래 글을 읽어보셔도 좋겠네요!
https://stackoverflow.com/questions/48817724/in-android-mvp-should-a-presenter-return-a-value

KwonDae replied:
프레젠터가 값을 return 하면 어떤 단점이 있을지 고민해보시고 공유해주세요!

Hevton replied:
[여기 내용](https://github.com/woowacourse/android-movie-theater/pull/66#discussion_r1589993290) 과 함께 여쭤보고 싶습니다!!

KwonDae commented on app/src/test/java/woowacourse/movie/SeatPresenterTest.kt:
`좌석 정보를 가져오면, ABC순서대로 정렬되어 있는 컬렉션이 존재한다` -> 해당 컬렉션을 검증하려면 어떻게 해야할까요?
`좌석을 클릭할 때 마다 가격은 10000원, 25000원 순서대로 가격을 반영시킨다` -> 순서대로 10000원, 25000원에 대해 presenter.calculateTotalPrice() 를 어떻게 검증할 수 있을까요?

사용하시는 mockk의 [Capturing](https://mockk.io/#capturing)의 CapturingSlot 과 MutableList를 활용해볼 수 있습니다.

KwonDae replied:
이부분은 어려울 수 있으니 진행하시면서 막히는 부분이 있다면 언제든 DM주세요!

Hevton replied:
감사합니다. Slot을 이용해서 컬러 테스트와 가격 테스트를 진행했습니다!!

KwonDae reviewed (changes requested):
팡태 안녕하세요!
미션 구현을 깔끔하게 구현해주셨습니다.

주로 Presenter, Adapter, Fragment 위주로 코멘트를 남겨드렸으니
확인 후 요청 부탁드립니다!

Commit:
:recycle: MovieDetailContract.kt 반환값 모두 삭제

Commit:
:recycle: onViewCreated() 다이어트

Commit:
:recycle: ktFormating

Commit:
:white_check_box: :recycle: 화면 이동 테스트 IntegrationTest.kt로 이동

Commit:
:sparkles: Intent오류 발생시 오류 Activity로 이동

Commit:
:sparkles: Activity, Fragment 바인딩기능 추가

Commit:
:recycle: Activity, Fragment 바인딩기능 적용

Commit:
:recycle: Cinema 프레젠터 내부로 이동

Commit:
:recycle: onClicked renameing

Commit:
:sparkles: Fragment redirectToErrorActivity

Commit:
:sparkles: TheatersBottomSheet MVP 구조 적용

Commit:
:recycle: MovieListPresenter.kt
loadMovie동작 프레젠터 내에서 init으로 적용

Commit:
:truck: showBottomSheet->navigateToCinemaView

Commit:
:truck: generateDateRange->loadDateRange

Commit:
:recycle: ktFormat

Commit:
:bug: init으로 옯긴 후 private val 위치 변경

Commit:
:recycle: 어댑터 구조 변경

Commit:
:recycle: 인텐트 전달 방식 변경

murjune committed:
:recycle: SeatPresenterTest capture 도입

murjune committed:
:recycle: fragment Container 이름 변경

murjune committed:
:recycle: timeDate formatting 개선

murjune committed:
:recycle: Time/Date Adapter 개선

murjune committed:
:recycle: ktlintCheck

murjune committed:
:truck: BindingActivity base package로 이동

murjune committed:
:recycle: ErrorActivity 시작 로직으로 변경

murjune committed:
:bug: Time/Date items MutableList로 변경

KwonDae commented on app/src/main/java/woowacourse/movie/movieList/MovieListFragment.kt:
```diff
@@ -54,7 +36,7 @@ class MovieListFragment : Fragment(), MovieListView {
+    override fun navigateToCinemaView(theater: Theater) {
         val fragment =
             (
                 childFragmentManager.findFragmentByTag(TheatersBottomSheetFragment.TAG)
```
https://github.com/woowacourse/android-movie-theater/pull/66#discussion_r1590036695

코멘트를 확인해주세요

KwonDae replied:
또한 바텀시트도 어떠한 데이터를 요구하는지 포맷을 제공해주도록 개선해볼 수 있을 것 같아요
THEATER_KEY 를 넘겨주어야 한다는 정보를 도메인 로직을 정확하게 알지 못하는 이상 내부 코드를 살펴보아야 합니다.

Hevton replied:
감사합니다!! 정말 유용한 방식인 것 같아요!

KwonDae commented on app/src/test/java/woowacourse/movie/SeatPresenterTest.kt:
```diff
@@ -1,86 +1,117 @@
-
+        val expectedColor = "#FF0000"
+        val colorSlot = slot<String>()
+        every { view.setSeatBackground(seatId, capture(colorSlot)) } just runs
+        // when
         presenter.toggleSeatSelection(seatId)
-        presenter.updateSeatBackground(seatId)
-
-        verify(exactly = 1) { presenter.toggleSeatSelection(seatId) }
-        verify(exactly = 1) { presenter.updateSeatBackground(seatId) }
+        // then
+        verify(exactly = 1) { view.setSeatBackground(seatId, colorSlot.captured) }
+        colorSlot.captured shouldBe expectedColor
```
mockk의 capturing을 잘 활용해주셨네요..!! 👍

Hevton replied:
알려주셔서 감사합니다!

KwonDae commented on app/src/test/java/woowacourse/movie/SeatPresenterTest.kt:
```diff
@@ -1,86 +1,117 @@
 
     @Test
-    fun `선택할 수 있는 좌석의 최대치는 티켓수 만큼이다`() {
+    fun `티켓수가 2일 때, 좌석은 2만큼 선택할 수 있다`() {
+        // given
         val seatId1 = "A1"
         val seatId2 = "A2"
+        val redColor = "#FF0000"
+        val seatOneColorSlot = slot<String>()
+        val seatTwoColorSlot = slot<String>()
+        every { view.setSeatBackground(seatId1, capture(seatOneColorSlot)) } just runs
+        every { view.setSeatBackground(seatId2, capture(seatTwoColorSlot)) } just runs
+        // when
+        presenter.toggleSeatSelection(seatId1)
+        presenter.toggleSeatSelection(seatId2)
+        // then
+        verify(exactly = 1) { view.setSeatBackground(seatId1, seatOneColorSlot.captured) }
+        verify(exactly = 1) { view.setSeatBackground(seatId2, seatTwoColorSlot.captured) }
+        seatOneColorSlot.captured shouldBe redColor
+        seatTwoColorSlot.captured shouldBe redColor
+    }
```
slot을 잘 활용해주셨는데요,
동일하게 [MutableList](https://mockk.io/#capturing) 를 활용한다면 두개의 슬롯 생성 없이도 테스트가 가능합니다.

Hevton replied:
아..!! MutableList는 두 개 이상이겠구나 했는데 맞군요! 감사합니다!!!

KwonDae commented on app/src/main/java/woowacourse/movie/movieList/MovieListPresenter.kt:
```diff
@@ -12,6 +12,12 @@ import woowacourse.movie.model.theater.Theater
 import java.time.LocalDate
 
 class MovieListPresenter(private val view: MovieListView) {
+    private val theaters: List<Theater> = generateTheaters(GENERATE_DUMMY_DATA_NUM)
```
MovieListPresenter를 보다 효과적으로 테스트 하기 위해선 해당 컬렉션을 생성자로 받으면 어떨까요?

Hevton replied:
감사합니다!! 그게 의존성 관리 측에서 더 좋은 방법이라고 생각됩니다!

KwonDae commented on app/src/main/java/woowacourse/movie/movieDetail/MovieDetailPresenter.kt:
```diff
@@ -1,43 +1,41 @@
     var ticketNum = 1
 
-    lateinit var movie: MovieInfo
+    val movie: MovieInfo = cinema.theater.movie
```
이부분도 MovieDetailPresenter 테스트를 위해 생성자로 관리할 수 있을까요?

Hevton replied:
넵! 감사합니다!!

KwonDae commented on app/src/main/java/woowacourse/movie/purchaseConfirmation/PurchaseConfirmationActivity.kt:
```diff
@@ -34,4 +40,26 @@ class PurchaseConfirmationActivity : AppCompatActivity() {
+        const val EXTRA_CINEMA = "cinema"
+        const val EXTRA_TIME_DATE = "timeDate"
+
+        fun newIntent(
+            context: Context,
+            ticketPrice: String,
+            seatNumber: Array<String>,
+            cinema: Cinema,
+            timeDate: String,
+        ): Intent {
+            return Intent(context, PurchaseConfirmationActivity::class.java).apply {
+                putExtra(EXTRA_TICKET_PRICE, ticketPrice)
+                putExtra(EXTRA_SEAT_NUMBER, seatNumber)
+                putExtra(EXTRA_CINEMA, cinema)
+                putExtra(EXTRA_TIME_DATE, timeDate)
+            }
+        }
```
어떤 값을 받아야 해당 Activity을 만들 수 있는지 명확해졌네요 👍

KwonDae commented on app/src/main/java/woowacourse/movie/ErrorActivity.kt:
```diff
@@ -0,0 +1,27 @@
+import android.view.MenuItem
+import androidx.appcompat.app.AppCompatActivity
+
+class ErrorActivity : AppCompatActivity() {
```
ErrorActivity가 HomeActivity와 같은 레벨에 있는게 다소 어색해보여요 🥲

Hevton replied:
인정입니다! 바꿨습니다!

KwonDae commented on app/src/main/java/woowacourse/movie/movieList/MovieAdapter.kt:
```diff
@@ -8,12 +8,12 @@ import androidx.recyclerview.widget.RecyclerView
+
     override fun getItemViewType(position: Int): Int {
         return if ((position + 1) % 4 == 0) ITEM_VIEW_TYPE_AD else ITEM_VIEW_TYPE_MOVIE
     }
```
하단의 `getItemCount` 함수와 이어지는 내용인데요,
Adapter에서 직접 itemCount 개수를 바꿔주는게 괜찮을까요?
예를 들어, size가 5인 리스트를 제공받았을 때

itemCount = 6 이므로 position은 5까지 늘어날 수 있습니다.
`RecyclerView`가 아이템을 그리면서 `onBindViewholder` 의 position이 5로 진입했을 때 어떤 일이 발생할까요?

KwonDae replied:
이 코멘트에 대한 힌트에 대해서 조금 얘기를 해보면

만일, 광고가 들어가는 로직이 복잡한 로직에 따라 결정이 된다면 Adapter 내에서 실행되어야 할거에요.
Adapter는 어떤 index에 광고가 위치해 있는지 관심이 있을까요?

```kotlin
sealed interface MovieDisplayData

data class Movie(..) : MovieDisplayData
data class Ad(..) : MovieDisplayData
```
데이터를 받을 때 `List< MovieDisplayData>` 로 받으면 복잡한 과정 없이 viewType을 지정해줄 수 있을거에요

Hevton replied:
맞습니다..! 이 부분 로직 모두 변경했습니다!

KwonDae reviewed (changes requested):
피드백 양이 많았는데 잘 반영해주셨네요 👍 
질문 주신 부분과 추가로 고민해볼만한 부분에 코멘트를 남겨드렸으니
확인 후 요청 부탁드립니다!

murjune committed:
:recycle: MovieDetailContract Presenter 네이밍 변경

- onTicketPlus -> increaseTicketCount
- onTicketMinus ->  decreaseTicketCount
- updateTimes -> updateRunMovieTimes
- loadDateRange -> loadRunMovieDateRange

Commit:
:truck: 바텀시트 파일 명확하게 변경

Commit:
:recycle: notifyDataSetChanged 추가

Commit:
:recycle: 불필요한 검증 제거

Commit:
:recycle: 데이터 넘겨주는 포맷 생성

Commit:
:recycle: slot -> MutableList

Commit:
:recycle: Presenter 생성자로 theaters 이동

Commit:
:truck: error 패키지 설정

Commit:
:recycle: 광고 처리 로직 변경

Commit:
:recycle: presenter 생성할때 theater바로 넘겨주기

Commit:
:recycle: onTheaterClicked 표시해주기!

Commit:
:recycle: adapter초기화 1회만

Commit:
:recycle: Presenter 비정상 종료시 finish되게 함

Commit:
:recycle: dataBinding 적용

Commit:
:recycle: dataBinding presenter 종속이 아닌 view만 알고있게 적용

Commit:
:recycle: ktLint Formating

Commit:
:recycle: Test File 함수변경에 따른 변경

Commit:
:bug: Test File 색 변경 잘못되어있는 부분 고침

murjune committed:
:recycle: TicketCountListener 로 네이밍 변경 및 패키지 분리

murjune committed:
:recycle: navigateToPurchaseConfirmation mvp 구조로 리팩토링

murjune committed:
:sparkles: MovieListMapper

murjune committed:
:sparkles: MovieDisplay UiModel 적용

murjune committed:
:recycle: SpinnerAdapter 객체 분리

- SpinnerDateAdapter
- SpinnerTimeAdapter

KwonDae reviewed (approved):
좋습니다 👍
1,2 단계에서 presenter와 adapter 관련해서 피드백 사항이 많았는데 잘 반영해주셨어요.
다음 단계 진행을 위해 머지하겠습니다.
이번 단계에서 마저 반영하지 못해 아쉬웠던 부분은 다음 단계에서 반영해주세요.
끝까지 파이팅입니다💪
